### PR TITLE
os/bluestore: bitmap allocator

### PR DESCRIPTION
**Pull request by chhabaramesh**
This is Bitmap based allocator for BlueStore.
Motivation:
1. Reduction of memory by using just one bit for single block (as compared to 8 + 8  byes in case of extent based allocator).
2. Concurrent allocations via multiple threads.

It exposes mainly these interfaces:
1. Allocate contiguous blocks.
2. Allocate non-contiguous blocks.
3. Free allocated blocks.
4. Get Statistics about allocator.

It works in two modes:

1. Concurrent:
 Where multiple threads can be allocating at same time in different space zones. This is specially for faster media like SSDs.
 
2. Serial:
 Only one thread can be active inside bitmap allocator at a time. This one is mainly to keep rotating media in mind since
 concurrent allocations can aggravate random writes.
 
Test Ran:
1. Make check.
2. Basic read write multi-threaded test cases with recovery.
3. Exhaustive unit test cases for corner cases

Next steps:
1. Complete set of multi threaded test cases.
2. More test cases to protect from space leak and corruption in BlueStore.
4. Provide more intelligent configurable like Zone size.
5. If required make scanning faster my one more level in summary hierarchy.
6. If required more stricter reservations guarantee.

Over All Design Overview:

Full space is divided in to fixed size but configurable zones. It is the unit where at a time single thread can be actively allocating from.
A zone is then further divided in to BitMaps. Each bitmap is 64bits hence 64 blocks. 

Allocation:

Start from a marker and scans for contiguous or dis-contiguous free blocks in zones one after another and in BitMaps one after another within a zone.
If while scan a zone is found locked by somebody, it simple tries in next zone.
If two scans does not find anything, then third scan is serial and it blocks on a zone if it is locked by another thread to avoid possibility missing only zones that have space.

In serial mode of allocation, it always take single lock on full allocator and scans for free space.

In contiguous allocation, the allocation is either fulfilled completely or nothing is allocated. 
In case of dis-contiguous allocations, partial results can be returned.

Free:
Free does not take any lock on zone or area, it simply atomically tweaks the bits.

Limitations:

1. Single allocation cannot extend beyond a zone in concurrent mode.

Signed-off-by: Ramesh Chander  <Ramesh.Chander@sandisk.com>

### CODE REVIEW

File: src/os/bluestore/Allocator.cc [hunk 1]
```diff
@@ -3,14 +3,29 @@
...
+  return NULL;
+}
+
+Allocator *Allocator::create(string type, int64_t size)
+{
+  if (type == "stupid") {
+    assert(0); // invalid call
+  } else if (type == "bitmap") {
+    return new BitMapAllocator(size);
+  }
```
Comment (liewegas) [inline src/os/bluestore/Allocator.cc]: Let's just make size a required argument and have a single create() factory.

File: src/os/bluestore/Allocator.h [hunk 2]
```diff
@@ -45,6 +45,7 @@ class Allocator {
   virtual void shutdown() = 0;
 
   static Allocator *create(string type);
+  static Allocator *create(string type, int64_t size);
```
Comment (liewegas) [inline src/os/bluestore/Allocator.h]: we can drop the size-less variant.

File: src/os/bluestore/BitAllocator.cc [hunk 3]
```diff
@@ -0,0 +1,1256 @@
...
+#include <assert.h>
+
+#define debug_assert assert
+#define MIN(x, y) ((x) > (y) ? (y) : (x))
+
+/*
+ * BmapEntityList functions.
+ */
+BmapEntityListIter::BmapEntityListIter(BmapEntityList *list) {
+	this->list = list;
```
Comment (liewegas) [inline src/os/bluestore/BitAllocator.cc]: style nit: { at start of next line in the .cc file (same line in .h file).

File: src/os/bluestore/BitAllocator.cc [hunk 4]
```diff
@@ -0,0 +1,1256 @@
...
+	this->start_idx = list->get_marker();
```
Comment (liewegas) [inline src/os/bluestore/BitAllocator.cc]: no need for this->

File: src/os/bluestore/BitAllocator.cc [hunk 5]
```diff
@@ -0,0 +1,1256 @@
+/*
```
Comment (liewegas) [inline src/os/bluestore/BitAllocator.cc]: please add the emacs/vi modelines at the top (top 2 lines of every other .cc or .h file in the repo).  they set tab stops at 2 chars to match the rest of the code base.

File: src/os/bluestore/BitAllocator.h [hunk 6]
```diff
@@ -0,0 +1,269 @@
+/*
+ * Bitmap based in memory allocator.
+ *
+ * Author: Ramesh Chander , Ramesh.Chander@sandisk.com
+ */
+
+#ifndef __BIT_ALLOCATOR_H
+#define __BIT_ALLOCATOR_H
```
Comment (liewegas) [inline src/os/bluestore/BitAllocator.h]: CEPH_OS_BLUESTORE_BITALLOCATOR_H

File: src/os/bluestore/BitAllocator.h [hunk 7]
```diff
@@ -0,0 +1,269 @@
...
+typedef enum bmap_alloc_mode {
+	SERIAL = 1,
+	CONCURRENT = 2,
+} bmap_alloc_mode_t;
+
+
+class ZoneList : public BmapEntityList {
+
+private:
+	pthread_mutex_t marker_mutex;
```
Comment (liewegas) [inline src/os/bluestore/BitAllocator.h]: let's use std::mutex

File: src/os/bluestore/BitMapAllocator.h [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+  int64_t num_uncommitted;
+  int64_t num_committing;
+  int64_t block_size;
+  int64_t num_reserved;
+
+  btree_interval_set<uint64_t> uncommitted; ///< released but not yet usable
+  btree_interval_set<uint64_t> committing;  ///< released but not yet usable
+  BitAllocator *bit_alloc; // Bit allocator instance
+
+  unsigned _choose_bin(uint64_t len);
```
Comment (liewegas) [inline src/os/bluestore/BitMapAllocator.h]: drop this

File: src/os/bluestore/BlueFS.cc [hunk 9]
```diff
@@ -280,9 +280,15 @@ void BlueFS::_init_alloc()
...
+    if (!bdev[id]) {
       continue;
-    alloc[id] = new StupidAllocator;
+    }
+    if (g_conf->bluestore_bmap_alloc) {
+	  assert(bdev[id]->get_size());
+	  alloc[id] = Allocator::create("bitmap", bdev[id]->get_size());
+    } else {
+	  alloc[id] = Allocator::create("stupid");
+    }
```
Comment (liewegas) [inline src/os/bluestore/BlueFS.cc]: this gets simpler with a single create()

File: src/os/bluestore/BlueStore.cc [hunk 10]
```diff
@@ -1044,7 +1044,12 @@ void BlueStore::_close_fm()
 int BlueStore::_open_alloc()
 {
   assert(alloc == NULL);
-  alloc = Allocator::create("stupid");
+  if (g_conf->bluestore_bmap_alloc) {
+	  assert(bdev->get_size());
+	  alloc = Allocator::create("bitmap", bdev->get_size());
+  } else {
+	  alloc = Allocator::create("stupid");
+  }
```
Comment (liewegas) [inline src/os/bluestore/BlueStore.cc]: ditto

File: src/test/objectstore/BitAllocator_test.cc [hunk 11]
```diff
@@ -0,0 +1,369 @@
+#include "os/bluestore/BitAllocator.h"
+#include <stdio.h>
+#include <assert.h>
+#define debug_assert assert
+
+void
+test_bmap_iter()
```
Comment (liewegas) [inline src/test/objectstore/BitAllocator_test.cc]: style nit: return value on same line

File: src/test/objectstore/BitAllocator_test.cc [hunk 12]
```diff
@@ -0,0 +1,369 @@
...
+{
+	int num_items = 5;
+	int off = 2;
+	class BmapEntityTmp : public BmapEntity {
```
Comment (liewegas) [inline src/test/objectstore/BitAllocator_test.cc]: indentation should be 2 spaces per level

File: src/os/Makefile.am [hunk 13]
```diff
@@ -46,6 +46,8 @@ libos_a_SOURCES += \
 	os/bluestore/ExtentFreelistManager.cc \
 	os/bluestore/FreelistManager.cc \
 	os/bluestore/KernelDevice.cc \
+	os/bluestore/BitMapAllocator.cc \
+	os/bluestore/BitAllocator.cc \
 	os/bluestore/StupidAllocator.cc
```
Comment (liewegas) [inline src/os/Makefile.am]: Please add the header files to the headers list in the same file.

There is also a cmake file that should be updated.

File: src/os/bluestore/BitAllocator.h [hunk 14]
```diff
@@ -0,0 +1,269 @@
...
+		return true;
+	}
+
+	virtual ~BmapEntity() { }
+};
+
+
+class BmapEntityList {
+	BmapEntity **items;
+	int64_t num_items;
```
Comment (liewegas) [inline src/os/bluestore/BitAllocator.h]: Can we use a standard type here (std::list or std::vector)?

File: src/os/bluestore/BitAllocator.h [hunk 15]
```diff
@@ -0,0 +1,269 @@
...
+	static uint64_t full_bmask();
+
+	static int64_t size();
+
+	static int empty_bmask();
+	static uint64_t align_mask(int x);
+	uint64_t bit_mask(int bit_num);
+
+	uint64_t
+	atomic_fetch() {return bits;}
```
Comment (liewegas) [inline src/os/bluestore/BitAllocator.h]: same line pls

File: src/os/bluestore/BitAllocator.h [hunk 16]
```diff
@@ -0,0 +1,269 @@
...
+
+private:
+	int64_t total_blocks;
+	int64_t zone_num;
+	int64_t used_blocks;
+
+	BmapEntityList *bmap_list;
+
+	enum {ZONE_FREE = 0, ZONE_ACTIVE = 1} state;
+	pthread_mutex_t lock;
```
Comment (liewegas) [inline src/os/bluestore/BitAllocator.h]: std::mutex

File: src/os/bluestore/Allocator.cc (hunk 1 above)
Comment (chhabaramesh) [inline src/os/bluestore/Allocator.cc]: Ok

File: src/os/bluestore/Allocator.h (hunk 2 above)
Comment (chhabaramesh) [inline src/os/bluestore/Allocator.h]: ok

File: src/os/bluestore/BitAllocator.cc (hunk 3 above)
Comment (chhabaramesh) [inline src/os/bluestore/BitAllocator.cc]: ok, will fix this and all other occurrences.

File: src/os/bluestore/BitAllocator.cc (hunk 5 above)
Comment (chhabaramesh) [inline src/os/bluestore/BitAllocator.cc]: ok

File: src/os/bluestore/BitAllocator.h (hunk 7 above)
Comment (chhabaramesh) [inline src/os/bluestore/BitAllocator.h]: ok

File: src/os/bluestore/BitMapAllocator.h (hunk 8 above)
Comment (chhabaramesh) [inline src/os/bluestore/BitMapAllocator.h]: ok

File: src/os/bluestore/BitAllocator.h (hunk 6 above)
Comment (chhabaramesh) [inline src/os/bluestore/BitAllocator.h]: ok

File: src/os/bluestore/BlueFS.cc (hunk 9 above)
Comment (chhabaramesh) [inline src/os/bluestore/BlueFS.cc]: ok

File: src/test/objectstore/BitAllocator_test.cc (hunk 11 above)
Comment (chhabaramesh) [inline src/test/objectstore/BitAllocator_test.cc]: ok. will change this and all other occurrences.

File: src/os/Makefile.am (hunk 13 above)
Comment (chhabaramesh) [inline src/os/Makefile.am]: ok

File: src/os/bluestore/BitAllocator.h (hunk 15 above)
Comment (chhabaramesh) [inline src/os/bluestore/BitAllocator.h]: ok

File: src/os/bluestore/BitAllocator.h (hunk 16 above)
Comment (chhabaramesh) [inline src/os/bluestore/BitAllocator.h]: ok

File: src/test/objectstore/BitAllocator_test.cc (hunk 12 above)
Comment (chhabaramesh) [inline src/test/objectstore/BitAllocator_test.cc]: done

File: src/os/bluestore/BlueStore.cc (hunk 10 above)
Comment (chhabaramesh) [inline src/os/bluestore/BlueStore.cc]: done

**Comment by liewegas** [comment]
also, please rebase on current master, as the Allocator::create() call in bluefs was fixed by another commit.

File: src/os/bluestore/BitAllocator.cc (hunk 4 above)
Comment (chhabaramesh) [inline src/os/bluestore/BitAllocator.cc]: I feel it makes it explicit and more readable. If not this we need some way of distinguish locals from class members. I can use any of those "m_*" ? or single underscore?

File: src/os/bluestore/BitAllocator.h (hunk 14 above)
Comment (chhabaramesh) [inline src/os/bluestore/BitAllocator.h]: these few lines are doing exactly what we need. I can tweak them or extend them the way I want. Given I am new to c++, I can start using STL as I feel more comfortable with it?

**Comment by liewegas** [comment]
./ceph_test_objectstore --gtest_filter=*/2

**Comment by liewegas** [comment]
CEPH_ARGS="--bluestore-freelist extent" ./ceph_test_objectstore ...

**Comment by liewegas** [comment]
https://github.com/ceph/ceph/pull/9152

**Pull request closed without merging**